feat: Add local gocryptfs support

daviewales opened the pull request:
Blocked by
- #1865
- #2039

This is a rebase of Germar Reitze's [gocrypt](https://github.com/bit-team/backintime/tree/gocrypt) branch on the latest `dev` branch.

It seems to work, so I thought we should try to merge it as a base for future gocryptfs work.

I've tested backing up a test file, deleting it, then restoring it.

Related to #1734.

Germar committed:
integrate gocrypt into Settings

Germar committed:
test gocryptfs on travis

Germar committed:
test gocryptfs on travis

Germar committed:
fix backintime-askpass not found

Germar committed:
fix gocryptfs stay in foreground

Germar committed:
fix gocryptfs stay in foreground

Germar committed:
reenable all tests

Germar committed:
reenable all tests

Germar committed:
set gocryptfs to be quiet

Germar committed:
set gocryptfs to be quiet

Germar committed:
typo

Germar committed:
typo

Germar committed:
fix settings

Germar committed:
reenable piping gocryptfs output as the bug is fixed in https://github.com/rfjakob/gocryptfs/issues/130

Germar committed:
reenable piping gocryptfs output as the bug is fixed in https://github.com/rfjakob/gocryptfs/issues/130

buhtz commented:
Wow that is a big step. Awesome!

Do we need the latest GoCryptFS when running TavisCI? Are there known Issues with gocryptfs from the Ubuntu repos?

daviewales commented:
I suspect that we don't need the latest version. I tested with the version in the latest Linux Mint, and it seemed to work.

daviewales commented:
The code fetching the latest version was already in Germar's original branch. I haven't made many changes here, other than fixing merge conflicts.

buhtz commented:
Please let me know if you prefer me to solve this merge conflicts.

daviewales commented:
I'll do another rebase, and switch Travis to fetch the packaged version of gocryptfs rather than the latest release while I'm at it.

daviewales committed:
Use gocryptfs from packagemanager in Travis

daviewales commented:
I've rebased on the latest `dev`.
I also added a quick patch to the `qt/configure` script to include the new `qt/manageprofiles` directory when installing.

I confirmed that I can still backup a file, delete it, then restore it.

buhtz commented on common/gocryptfstools.py:
```diff
@@ -0,0 +1,135 @@
+#
+#    You should have received a copy of the GNU General Public License along
+#    with this program; if not, write to the Free Software Foundation, Inc.,
+#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
```
```suggestion
# SPDX-FileCopyrightText: © 2017 Germar Reitze
#
# SPDX-License-Identifier: GPL-2.0-or-later
#
# This file is part of the program "Back In time" which is released under GNU
# General Public License v2 (GPLv2). See LICENSES directory or go to
# <https://spdx.org/licenses/GPL-2.0-or-later.html>.
```

buhtz commented on common/gocryptfstools.py:
```diff
@@ -0,0 +1,135 @@
+from mount import MountControl
+from exceptions import MountException
+
+_ = gettext
```
```suggestion
```

No need for this because we do "install gettext" on a global level. Don't know the correct term now. But in the end there is no need to import gettext or declare _.
This can cause problems in our unit tests because they do not follow modern packaging standards yet.
For example see common/tools.py
```
# While unittesting and without regular invocation of BIT the GNU gettext
# class-based API isn't setup yet.
try:
    _('Warning')
except NameError:
    _ = lambda val: val
```

buhtz commented on common/gocryptfstools.py:
```diff
@@ -0,0 +1,135 @@
+
+import os
+import subprocess
+from gettext import gettext
```
```suggestion
```

Please see my next comment.

buhtz commented on qt/configure:
```diff
@@ -162,6 +162,8 @@ addUninstallDir          "/share/backintime/qt/__pycache__"
 addInstallDir            "/share/backintime/qt"
 addInstallFiles "*.py"   "/share/backintime/qt"
+addInstallDir            "/share/backintime/qt/manageprofiles"
+addInstallFiles "manageprofiles/*.py"   "/share/backintime/qt/manageprofiles"
```
Great. Thank you.
This was my lazy testing.

buhtz replied:
Add this to PR #1900 .

buhtz commented:
When my refactoring work on the manage profiles dialog is finished I think about a release candidate and release 1.5.3.
After this we can merge your PR, after intense testing of course, and later release it as 1.6.0 somewhere around Christmas or New Year.

buhtz commented:
FYI: I do plan a [1.5.3 release](https://github.com/bit-team/backintime/milestone/31) in the near future. After that release I would target your PR for a [1.6.0 release](https://github.com/bit-team/backintime/milestone/33).

daviewales commented:
I'll apply the suggestions above when I get a chance. There are also a few lint errors, so I'll fix those up too.

daviewales committed:
Update common/gocryptfstools.py

Co-authored-by: buhtz <c.buhtz@posteo.jp>

daviewales committed:
Remove explicit gettext definition

Co-authored-by: buhtz <c.buhtz@posteo.jp>

daviewales committed:
Remove explicit gettext definition

Co-authored-by: buhtz <c.buhtz@posteo.jp>

daviewales committed:
Fix some pylint warnings

- Remove unused import
- Explicitly declare class members

daviewales commented:
The remaining lint errors are because there is similarity between the encfs and gocryptfs modules... Perhaps this can be ignored?

buhtz commented:
> The remaining lint errors are because there is similarity between the encfs and gocryptfs modules... Perhaps this can be ignored?

Usually I would say "no" and recommend to create a base class to avoid code duplication. But we will remove encfs, so I see no problem to add an ignore clause for pylint.

    # pylint: disable=duplicate-code

Not sure if this need to be place in both files.

daviewales committed:
Disable pylint duplicate check for encfs module

- It's OK to remove these checks for encfs, as we plan to deprecate and
  remove the encfs code.

daviewales committed:
Fix  undefined variable

daviewales commented:
Note to self: I need to test setting up a new `gocryptfs` backup from scratch. I've been tested this branch with one which is already configured.

EDIT: I've tested setting up a new backup profile from scratch, and the dialog needs some tweaking as it's missing a few inputs. I'll update it to match the EncFS local dialog.

buhtz commented:
Hello David,
your PR is the next. I leave it up to you to update your branch to the latest dev version.
Then I will do some manual testing in a VM.

Christian

daviewales commented:
Thanks Christian,
I'll try to find some time in the next week to get this ready. If you have a chance, you might be able to help me figure out why the new GoCryptFS profile creation screen is missing a few important fields compared to the old EncFS local screen. Otherwise, I'll see if I can figure it out.

buhtz commented on common/gocryptfstools.py:
```diff
@@ -0,0 +1,129 @@
+from exceptions import MountException
+
+
+class GoCryptFS_mount(MountControl):
```
I asked at the gocryptfs project how their name is officially spelled. It is spelled with all letters in lower case.

    gocryptfs

Because of that and PEP8 I would suggest the class name `GocryptfsMount`.

daviewales replied:
Done.

buhtz commented:
> why the new GoCryptFS profile creation screen is missing a few important fields compared to the old EncFS local screen. Otherwise, I'll see if I can figure it out.

Not sure. Can you show me a screenshot?
I am not able to use your current branch because of merge conflicts. I assume they are caused because of my refactoring of the manage profiles dialog.

buhtz commented:
Hello David,
I will try this weekend to take a closer look to your PR. Don't worry if it isn't finished yet.
But before please take care that the PR is up-2-date with your local branch and the latest dev version.

Regards,
Christian

daviewales committed:
Add more gocryptfs details

daviewales committed:
Standardise GocryptfsMount class name

daviewales commented:
**EDIT**: I've fixed this in e8a66fa485a97b4643cd5dc1eebb67bdbed6c2ac.

When I create a new EncFS local profile, I see these fields:

![image](https://github.com/user-attachments/assets/7bf452c0-75f2-445e-81d9-f3ad4a27859d)

But when I create a new gocryptfs local profile, I see these fields:

![image](https://github.com/user-attachments/assets/36cb23ba-52f8-450f-b420-4455988457f2)

i.e. the _Where to save snapshots_ field is missing.

Do you know how to make it appear?

daviewales commented:
Also, I've rebased on the latest `dev`.

daviewales committed:
Ensure GUI elements appear for local_gocryptfs mode

daviewales commented:
I'm getting the following error when I click 'OK' to finalise creation of a new profile:

```
DEBUG: [common/config.py:320 Config.checkConfig] Check profile gocryptfs-local
Traceback (most recent call last):
  File "/usr/share/backintime/qt/manageprofiles/__init__.py", line 794, in accept
    if self.validate():
       ^^^^^^^^^^^^^^^
  File "/usr/share/backintime/qt/manageprofiles/__init__.py", line 553, in validate
    if not self.config.setupCron():
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/share/backintime/common/config.py", line 1468, in setupCron
    self.setupUdev.clean()
  File "/usr/share/backintime/common/tools.py", line 2441, in clean
    self.iface.clean()
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 634, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.AccessDenied: Rejected send message, 1 matched rules; type="method_call", sender=":1.201" (uid=1000 pid=100353 comm="/usr/bin/python3 -Es /usr/share/backintime/qt/app." label="unconfined") interface="net.launchpad.backintime.serviceHelper.UdevRules" member="clean" error name="(unset)" requested_reply="0" destination=":1.88" (uid=0 pid=16206 comm="/usr/bin/python3 -Es /usr/share/backintime/qt/serv" label="unconfined")
```

Do you have any thoughts?

buhtz committed:
fix errorHandler

buhtz commented:
Mhm... I am not able to reproduce your dbus related error. But I do see something like this.
![image](https://github.com/user-attachments/assets/ba26e81b-5fc3-4d40-904d-9ba8d8bd8410)

So I executed the gocryptfs command from the error message directly in the shell:
```sh
$ gocryptfs -extpass backintime-askpass -quiet /home/user/Dokumente /home/user/.local/share/backintime/mnt/1CD92332/mountpoint
WARNING: Failed to connect to Udev serviceHelper daemon via D-Bus: org.freedesktop.DBus.Error.Spawn.ChildExited
WARNING: D-Bus message: Launch helper exited with unknown return code 1
WARNING: Udev-based profiles cannot be changed or checked due to Udev serviceHelper connection failure
extpass: password is empty
```

"password is empty" is the relevant message.

**EDIT**: Currently I have no idea.

**EDIT2**:  I am assuming for someone who understands DBus your error message would be easy to understand. But for me. ;) What kind of system do you use?

> `AccessDenied: Rejected send message`

?

> `interface="net.launchpad.backintime.serviceHelper.UdevRules"`

Did you tried to setup udev scheduling?

> `comm="/usr/bin/python3 -Es /usr/share/backintime/qt/serv" label="unconfined")`

That is interesting. The file `/usr/share/backintime/qt/serv` does not exist on my system.

**EDIT3**: I am assuming we do have two separate problems here. On your system something doesn't work related to DBus. On my system something is wrong with backintime-askpass. Further I am assuming that my problem will occur on your system too if you fix your DBus problem.

daviewales commented:
I have another gocryptfs profile which I created in another experimental branch. That was working before my latest rebase. Now it's stopped working with a similar `extpass: password is empty` message. So something strange is happening with `backintime-askpass`.

buhtz commented:
That is a trail. If you can identify a commit where it worked, we can do a diff.

daviewales commented:
I can also confirm that the DBusException above occurs even on the latest `dev`, with a completely new `local` profile.

I've run `git bisect`, and for each bisection I've deleted the config, and created a new local profile from scratch.
This appears to be the first commit with the issue, but I have no idea why!
59746799b31a853b071f22fe67501e315d3e2633

daviewales commented:
Another experiment, `git revert 59746799b31a853b071f22fe67501e315d3e2633` on this branch makes everything work!

![image](https://github.com/user-attachments/assets/be7452a3-f56b-4049-bf1b-fac200c00853)

daviewales commented:
It looks like adding the license information comment to `net.launchpad.backintime.serviceHelper.conf` broke DBus!

daviewales commented:
I've sent a PR to fix this: #1958

daviewales committed:
Fix typo in docstring

buhtz commented:
I still get this askpass error I described. Can you reproduce it?

daviewales commented:
I'm not seeing any errors on this branch now. Could you try running with a fresh config file? How exactly are you running it?

daviewales commented:
What does the output from `--debug` look like just before you get that error? Which secret storage backend does the log indicate you are using?

buhtz commented:
I do run this on a VM with Debian 12. This is the full debug output.

```
user@worf ~/backtinime_daviewales/qt (gocrypt-rebase)$ ./backintime-qt --debug
DEBUG: [common/backintime.py:616 argParse] Argument(s) used: {'debug': True, 'quiet': False}
DEBUG: [common/backintime.py:515 startApp] backintime: {'name': 'Back In Time', 'version': '1.6.0-dev.506b31c6', 'running-as-root': False}; host-setup: {'OS': {'/etc/os-release': 'Debian GNU/Linux 12 (bookworm)', '/etc/debian_version': '12.8\n'}}

Back In Time
Version: 1.6.0-dev.506b31c6

Back In Time comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type `backintime --license' for details.

DEBUG: [common/configfile.py:591 Config.setCurrentProfile] Change current profile: 1=Hauptprofil
WARNING: [common/tools.py:2382 __init__] Failed to connect to Udev serviceHelper daemon via D-Bus: org.freedesktop.DBus.Error.Spawn.ChildExited
WARNING: [common/tools.py:2384 __init__] D-Bus message: Launch helper exited with unknown return code 1
WARNING: [common/tools.py:2385 __init__] Udev-based profiles cannot be changed or checked due to Udev serviceHelper connection failure
DEBUG: [common/tools.py:226 initiate_translation] No language code. Use systems current locale.
DEBUG: [common/tools.py:241 initiate_translation] Language code used: "de"
DEBUG: [common/backintime.py:698 getConfig] config file: "/home/user/.config/backintime/config"; share path: "/home/user/.local/share/backintime"; profiles: "1=Hauptprofil"
DEBUG: [common/pluginmanager.py:233 PluginManager.load] Register plugin path /home/user/backtinime_daviewales/common/plugins
DEBUG: [common/pluginmanager.py:233 PluginManager.load] Register plugin path /home/user/backtinime_daviewales/qt/plugins
DEBUG: [common/tools.py:1084 is_Qt_working] Qt probing result: exit code 2
DEBUG: [common/tools.py:1087 is_Qt_working] Qt probing stdout:

DEBUG: [common/tools.py:1088 is_Qt_working] Qt probing errout:
DEBUG: [common/qt_probing.py:89 <module>] /home/user/backtinime_daviewales/common/qt_probing.py started... Call args: ['/home/user/backtinime_daviewales/common/qt_probing.py', '--debug']
DEBUG: [common/qt_probing.py:90 <module>] Display system: x11
DEBUG: [common/qt_probing.py:91 <module>] XDG_RUNTIME_DIR=/run/user/1000
DEBUG: [common/qt_probing.py:92 <module>] XAUTHORITY=/home/user/.Xauthority
DEBUG: [common/qt_probing.py:93 <module>] QT_QPA_PLATFORM=($QT_QPA_PLATFORM is not set)
DEBUG: [common/qt_probing.py:95 <module>] Current euid: 1000
DEBUG: [common/qt_probing.py:127 <module>] isSystemTrayAvailable for Qt: True
DEBUG: [common/qt_probing.py:132 <module>] /home/user/backtinime_daviewales/common/qt_probing.py is terminating normally (exit code: 2)

DEBUG: [plugins/systrayiconplugin.py:74 init] System tray is available to show the BiT system tray icon
DEBUG: [common/pluginmanager.py:254 PluginManager.load] Add plugin systrayiconplugin.py
DEBUG: [common/pluginmanager.py:254 PluginManager.load] Add plugin notifyplugin.py
DEBUG: [qt/qttools.py:280 createQApplication] QT QPA platform plugin: xcb
DEBUG: [qt/qttools.py:281 createQApplication] QT_QPA_PLATFORMTHEME=<not set>
DEBUG: [qt/qttools.py:286 createQApplication] QT_STYLE_OVERRIDE=<not set>
DEBUG: [qt/qttools.py:289 createQApplication] QT active style: fusion
DEBUG: [qt/qttools.py:290 createQApplication] QT fallback style: 
DEBUG: [qt/qttools.py:291 createQApplication] QT supported styles: ['Windows', 'Fusion']
DEBUG: [qt/qttools.py:292 createQApplication] themeSearchPaths: ['/usr/share/icons', ':/icons']
DEBUG: [qt/qttools.py:293 createQApplication] fallbackSearchPaths: []
DEBUG: [qt/qttools.py:298 createQApplication] Is SystemTray available: True
DEBUG: [qt/qttools.py:320 createQApplication] Trying to set App ID for non-privileged user
DEBUG: [qt/qttools.py:360 initiate_translator] No language code. Use systems current locale.
DEBUG: [qt/icon.py:20 <module>] Checking if the current theme contains the BiT icon...
DEBUG: [qt/icon.py:32 <module>] Found an installed theme: Tango
DEBUG: [common/config.py:1413 Config.isConfigured] Profile (profile_id=None) is not configured because snapshot path is "False" and/or includes are "False".
DEBUG: [manageprofiles/tab_general.py:583 _snapshot_mode_combobox] snapshot_modes={'local': 'Lokal', 'ssh': 'SSH', 'local_encfs': 'Lokal verschlüsselt (EncFS)', 'ssh_encfs': 'SSH verschlüsselt', 'local_gocryptfs': 'Lokal verschlüsselt (gocryptfs)'}
DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/config.py:1413 Config.isConfigured] Profile (profile_id='1') is not configured because snapshot path is "False" and/or includes are "False".
DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/config.py:1413 Config.isConfigured] Profile (profile_id=None) is not configured because snapshot path is "False" and/or includes are "False".
DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/gocryptfstools.py:128 GocryptfsMount.isConfigured] No config in /home/user/backup_dest/gocryptfs.conf
DEBUG: [common/gocryptfstools.py:76 GocryptfsMount.init] Provide password through temp FIFO
DEBUG: [common/gocryptfstools.py:85 GocryptfsMount.init] Call command to create gocryptfs config file: gocryptfs -extpass backintime-askpass -init /home/user/backup_dest

...ERROR DIALOG appears here....

DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/gocryptfstools.py:128 GocryptfsMount.isConfigured] No config in /home/user/backup_dest/gocryptfs.conf
DEBUG: [common/gocryptfstools.py:76 GocryptfsMount.init] Provide password through temp FIFO
DEBUG: [common/gocryptfstools.py:85 GocryptfsMount.init] Call command to create gocryptfs config file: gocryptfs -extpass backintime-askpass -init /home/user/backup_dest
DEBUG: [common/tools.py:1475 keyringSupported] Keyring config file directory: /home/user/.config/python_keyring
DEBUG: [common/tools.py:1490 keyringSupported] Available keyring backends:
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.libsecret.Keyring (priority: 4.8)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.SecretService.Keyring (priority: 5)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.fail.Keyring (priority: 0)
DEBUG: [common/tools.py:1494 keyringSupported] keyring.backends.chainer.ChainerBackend (priority: 10)
DEBUG: [common/tools.py:1550 keyringSupported] Not found Metaclasses: ['keyring.backends.Gnome.Keyring', 'keyring.backends.kwallet.Keyring', 'keyring.backend.SecretServiceKeyring', 'keyring.backend.GnomeKeyring', 'keyring.backend.KDEWallet']
DEBUG: [common/tools.py:1551 keyringSupported] Available supported backends: [<class 'keyring.backends.SecretService.Keyring'>, <class 'keyring.backends.kwallet.DBusKeyring'>, <class 'keyring.backends.chainer.ChainerBackend'>]
DEBUG: [common/tools.py:1554 keyringSupported] Found appropriate keyring 'keyring.backends.chainer'
DEBUG: [common/gocryptfstools.py:128 GocryptfsMount.isConfigured] No config in /home/user/backup_dest/gocryptfs.conf
DEBUG: [common/gocryptfstools.py:76 GocryptfsMount.init] Provide password through temp FIFO
DEBUG: [common/gocryptfstools.py:85 GocryptfsMount.init] Call command to create gocryptfs config file: gocryptfs -extpass backintime-askpass -init /home/user/backup_dest
DEBUG: [common/configfile.py:591 Config.setCurrentProfile] Change current profile: 1=Hauptprofil
DEBUG: [common/config.py:1413 Config.isConfigured] Profile (profile_id=None) is not configured because snapshot path is "False" and/or includes are "False".
DEBUG: [common/config.py:1413 Config.isConfigured] Profile (profile_id=None) is not configured because snapshot path is "False" and/or includes are "False".
```

I am not able to finishe the profile creation because of this error dialog.

![image](https://github.com/user-attachments/assets/7afb9879-b79c-4d82-96b5-2776f2102df2)

daviewales commented:
Thanks. Do you get that error when creating a new profile of another type? (Local or Local EncFS)

I also notice one difference, which is that I install the program rather than running from the source folder directly.

daviewales commented:
I've tested, and I get a similar error if I launch `./backintime-qt --debug` from the source directory, but not if I install it then run it.

This gives the error:

```
~/backintime/qt$ ./backintime-qt --debug
```

![image](https://github.com/user-attachments/assets/3c1ac2cf-0682-4137-a9c9-d04185e77d81)

But this works without error:

```
cd ~/backintime/qt
./configure && make && sudo make install
cd ../common
./configure && make && sudo make install
backintime-qt --debug
```

daviewales commented:
While testing the above, I discovered a regression for other snapshot methods. This PR adds a 'high-level init' method to `mount.Mount`. In turn, this expects sub-classes which require mounting to define their own `init` method. However, we have not defined this for other sub-classes, so they currently fail.

buhtz commented:
> Do you get that error when creating a new profile of another type? (Local or Local EncFS)
When creating a fresh and new profile. I started BIT the first time without having an existing config file.

I installed BIT but then (out of habit) started it from the source folder. I'll try later to run the installed version.

> This PR adds a 'high-level init' method to mount.Mount

Do you mean `mount.py::Mount.init()`? I would suggest to rename it to something else because "init" is kind of reserved in its meaning and also to broad. I don't have a good idea. "high_level_init()" might not be nice but OK to me. It don't have to "sound good". ;)

buhtz commented:
OK, I can confirm now that it works in a fresh and clean install. Don't know what was interfering. It was not only the way how I started BIT. I needed to rollback my VM to an earlier and virgin state and re install BIT. Then it worked.

daviewales commented:
I think an `init` method is added in two places. I've brought it across directly from Germar's original branch. I'll try to think of an alternative name.

I need to look it more deeply, as it appears that Germar was intending to slightly redesign the Mount API to require this method for other instances such as Local EncFS too.

daviewales committed:
Rename init to init_backend

To avoid confusion with __init__

daviewales commented:
I've renamed `init` to `init_backend` and rebased on `dev`. I haven't tested it because I'm hitting this issue: https://github.com/bit-team/backintime/pull/1996.

buhtz commented:
> I'm hitting this issue: #1996.

I merged it now.

daviewales commented:
The gocryptfs functionality works. But Germar changed the `mount` API slightly to require mount backends to include an `init_backend()` method. (Was originally `init()`, but I changed it to avoid confusion.)

However, the branch I rebased never updated the encfs mount backends to include this method.

From my initial inspection, I think that `init_backend` is very similar to the existing `_mount` method, but it extracts the code required to setup a new backend if it is not yet configured.

buhtz commented:
Great! You can add your name to the header of the files (see: `SPDX-FileCopyrightText:`). Decide yourself if you provide a nickname or your real name. Using email might cause spam.

I promoted that PR
- at Mastodon: https://fosstodon.org/@backintime/113803972815472192
- at gocryptfs community itself: https://github.com/rfjakob/gocryptfs/discussions/889

daviewales commented:
Thanks. I'll try to find time over the next week or two to update the EncFS module to use the new `init_backend` method. After that, I think it might be ready.

daviewales committed:
Slightly improve code style

daviewales committed:
Add init_backend method to match gocryptfs updates

We are rebasing previous gocryptfs work from 2017. This work added an
`init_backend` method which  separates the concerns of first-time
initialisation vs subsequent use.

daviewales commented:
I've updated EncFS with an `init_backend` method to match the new Mount structure in the gocryptfs branch.
It seems to work.

I've noticed one more missing feature.
EncFS would backup the EncFS configuration into the local config folder.
But currently, we don't have an equivalent `backupConfig` method for `gocryptfs`.
It should be pretty straightforward to copy the EncFS implementation so I'll try that soon.

**EDIT**: There's still a bug with EncFS. I'll see if I can track it down.

buhtz commented:
> **EDIT**: There's still a bug with EncFS. I'll see if I can track it down.

In this case I recommend to set the PR into Draft mode.

If you describe the bug others might be able to help.

daviewales commented:
In the branch which I've rebased, it defines a new method for MountControl, which explicitly initialises the backend. In the original branch this was called `init`, but I've renamed it to `init_backend` to avoid confusion.

The new Gocryptfs code in this branch works with this refactored MountControl API, but the EncFS code was never updated to use it. In my latest commits, I've tried to adapt the EncFS code to use the explicit `init_backend` method. However, there is a bug such that when you create a new EncFS profile, somehow the mount is not correctly created, and an exception is raised.

I would appreciate your thoughts about whether we should really be changing the MountControl API in this PR, or whether we should try to split these changes. For example if you agree that the API change to require explicitly initialising the backend is a good idea, perhaps we could try to get just that change working and merged first, then come back to this Gocryptfs work.

On the other hand, if you feel that this `init_backend` change is a bad idea, then I can rework this branch to avoid making any changes to the MountControl API.

buhtz commented:
Hello David,

thank you for the summary.

> I would appreciate your thoughts about whether we should really be changing the MountControl API in this PR, or whether we should try to split these changes. For example if you agree that the API change to require explicitly initialising the backend is a good idea, perhaps we could try to get just that change working and merged first, then come back to this Gocryptfs work.

Splitting the changes sounds the most elegant to me. First refactor/fix the mount API please.

Regards,
Christian

buhtz commented:
Just FYI. Debian announced their schedule for the next release. The two important dates are the 15th April (Soft Freeze) and 15th May (Hard Freeze). These dates are not written in stone and it is not uncommon for these to be postponed even further.

Anyway we should try to get the next upstream release of BIT done until 1st April, which is two weeks before the Soft Freeze. In this case our Debian package maintainer has two weeks to finalize the packaging.

Before the release, I would like to make a release candidate for a period as long as possible. But if there is not enough time for this, so be it.

Under circumstances and with friendly Debian maintainers it might be possible for us to bring a new BIT release into Debian even after the Soft Freeze, but before the Hard Freeze. So there might be one month more for us. The EncFS removal should be a good argument, related to security, to propose this.

Please don't get me wrong. This is a FOSS project with volunteers only. No one get hurt or fired if we don't reach the deadlines. So be it. We do our best with respect to our self and our private life.

buhtz commented:
Note to myself: Adapt man page to new situation.

daviewales commented:
No guarantees for the April deadline, but I'll see what I can do. I'm trying a more cautious incremental approach to avoid the issues above. My plan is to submit the following changes as three separate PRs:

- [ ] Create tests for `encfstools` (WIP: https://github.com/bit-team/backintime/pull/2039)
- [ ] Update mount API with explicit `init_backend` method (and ensure it passes the `encfstools` tests)
- [ ] Add `gocryptfs` support

buhtz commented:
Hello David,
the 1.5.4 release is published. Therefore, the road is free now.

I don't have the overview currently. To my understanding everything related to gocrypt is work in progress. Am I right?

Let me know if I should take a look on something or test something. I usually read all notifications from Microsoft GitHub. But feel free to contact me via email or mailing list. Your task has priority now.

Regards,
Christian

daviewales commented:
Thanks @buhtz. Everything is WIP currently. The current priority is to get some good tests for the existing EncFS and MountControl APIs. This is important, as this gocryptfs branch, which was rebased from Germar's old branch, changes the MountControl API to add an explicit `init_backend` method. However, this broke the existing EncFS module, and made me feel that we need to strongly test the current working state before changing the MountControl API.

I haven't had much bandwidth lately to look at this, but I do hope to get back into it when I get a chance.
If you have a chance to add some tests to #2039, it would be very helpful.

buhtz commented:
Some more details in this comment: https://github.com/bit-team/backintime/pull/2039#issuecomment-3472633268

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
[skip ci]

buhtz committed:
[skip ci]

buhtz commented:
Hello David,
I just made some minor adjustments and fixes in the GUI code but not to your gocryptfs.

I still lack of deeper understanding **but** after a first test, the local gocryptfs-encrypted backup seems to work.

Looks better than I thought. Did you remember any other important problems with the current state, we need to tackle before merging?

Of course, I'll review the modifications line by line before merging anything.

Regards,
Christian

daviewales commented:
@buhtz, the main problem with this branch is that it breaks EncFS, because it changes the `MountControl` API to require an `init_backend()` method, which is explicitly called when needed. This conflicts with the existing `MountControl` API, which handles initialisation inline.

Some details above:
https://github.com/bit-team/backintime/pull/1897#issuecomment-2629038425

Previously, I was trying to update the EncFS code to use this new API.
However, I now feel that it's probably better to leave the EncFS code unchanged, and update this branch to work with the existing `MountControl` api. (See the current EncFS code for an example of how this would work.)

buhtz commented:
> it breaks EncFS

Ah, stupid me. :smile: Have not tested this case. ;)
Thanks for pointing me to it. :heart:

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz commented:
Mhm.... OK, do I understand that right? The `init_backend()` is called only when creating a profile to initialize/setup the encrypted directory?

I checked the EncFS related code. In the current version there is an `init_backend()` added in this PR. This method does not exist in the latest `dev` version, but EncFS profiles do work in `dev`. Why? I am assuming that also an encfs profile somehow need to initialize its destination folder. But where in the `dev` code (without having `init_backend()`) does this happen?

--- 

**EDIT**: Not sure about it, but I am assuming it happens in `Encfs_mount._mount()` kind of implicit with this two lines:

            if not self.isConfigured():
                encfs += ['--standard']

The `--standard` switch makes EncFS initializing the directory before doing something else.

The `Encfs_mount.init_backend()` in this PR looks like a near copy of `Encfs_mount._mount()`.

Why not just leave `Encfs_mount.init_backend()` empty?

At the current point I am not interested in fancy and high level super duper mega developer designs. I just want to make it work. :rofl: And EncFS is on the way to be removed soon. So I have no problems keeping some awkward workarounds until its final removal.

buhtz committed:
Encfs_mount.init_backend() just empty

buhtz committed:
Merge branch 'gocrypt-rebase' of github.com:daviewales/backintime into gocrypt-rebase

daviewales commented:
Yes, I think most (all?) of the EncFS backend init logic is in these lines:
https://github.com/bit-team/backintime/blob/a8dcbff133fa4d1f7d67468d6917adcb146f18a2/common/encfstools.py#L77-L78

Note that the `isConfigured` method does more than just check if it's configured. It also confirms through the user interface whether the user wants to create a new EncFS config, and if so, asks them to confirm the password.
It can also raise a few exceptions for various reasons.

I have a feeling that I tried just leaving `init_backend()` empty, but I can't remember whether it worked, or if there was some problem. Might be worth trying that again.

buhtz commented:
Thank you for your feedback. OK, I'll try to leave it empty.

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
encfs fifo docu

buhtz committed:
more doc and encfs empty initbackend

buhtz committed:
pepification

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz commented:
I tend to simply merge it.
Still don't feel sure and calm about it because it is not my area of expertise. But I also see no clear reasons why keep it pending.

buhtz committed:
approach translator cutoff from 99 to 97 [skip ci]

buhtz committed:
minor adjustments

buhtz committed:
Merge branch 'dev' into gocrypt-rebase

buhtz committed:
lint fix

buhtz committed:
changelog and typos [skip ci]

buhtz commented:
Hello David,
I have one further question to make sure I am on the right track.

You or Germar introduced `Mount.isConfigured()`. This cause #2360 because `class SSH` does not implement that method.

In the past only `EnCFS_mount.isConfigured()` existed. Do you see a good reason to keep `Mount.isConfigured()`?

My current approach would be pragmatic. I'll would like to remove it from `class Mount` and just keep it exclusive for `class EncFS_mount` and `class GocryptfsMount`.

Re-designen the whole mounting stuff is something I wouldn't like to do now.

What do you think?

Regards,
Christian

daviewales commented:
@buhtz, yes, I support removing `Mount.isConfigured()` from the high-level API, to avoid breaking other mount types. I think it was originally added in [Germar's branch](https://github.com/bit-team/backintime/blob/2d26794f06ba661a5bedb614c9fb4954a6b7550c/common/mount.py#L383-L411), but that history seems to have gone missing from the merged commit.

Originally, I was trying to update all the other mount types to use the new methods from Germar's gocrypt branch, but then I changed direction, and thought it would be better to leave the mount API unchanged. That's when I went down the rabbit hole of trying to write tests for the different mount types.

In summary, I agree that it's a mistake to change the Mount API in ways that break other mount types.
Instead, as you propose, we should just make the gocryptfs mount work with the original API.
